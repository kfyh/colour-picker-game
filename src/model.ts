export const STATES = {
  STARTED: 'STARTED',
  SELECTING: 'SELECTING',
  CYCLING: 'CYCLING',
  SHOWRESULT: 'SHOWRESULT',
  ENDED: 'ENDED',
};

export interface Model {
  state: {
    currentResultIndex: number;
    nextResultIndex: number;
    currentUserSelectedIndex: number;
    currentState: string;
  };
  settings: {
    colours: number[];
  };
}
