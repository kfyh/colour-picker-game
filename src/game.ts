import { Model, STATES } from './model';
import { randomInt } from './utils/randomInt';

export class Game {
  private readonly model: Model;
  private time: number;
  constructor(model: Model) {
    this.model = model;
    this.time = 0;
  }

  public setSelected(index: number): void {
    if (
      this.model.state.currentState === STATES.SELECTING ||
      this.model.state.currentState === STATES.STARTED
    ) {
      this.model.state.currentUserSelectedIndex = index;

      this.model.state.currentResultIndex = -1
      this.model.state.currentState = STATES.CYCLING;
      this.model.ui.roundResult = '';
      this.time = 0;

      randomInt(0,4)
        .then((result) => {
          this.model.state.currentResultIndex = result;
        }, null);
    }
  }

  public update(delta: number): void {
    this.time += delta;
    if (
      this.model.state.currentState !== STATES.STARTED &&
      this.model.state.currentState !== STATES.ENDED
    ) {
      this.model.state.timeRemaining -= delta;
      if (this.model.state.timeRemaining <= 0) {
        this.model.ui.roundResult = "Time's Up!";
        this.model.state.currentState = STATES.ENDED;
      }
    }

    if (this.model.state.currentState === STATES.CYCLING) {
      if (this.time >= 2000 && this.model.state.currentResultIndex >= 0) {
        this.model.state.currentState = STATES.SHOWRESULT;
        this.time = 0;
      }
    }

    if (this.model.state.currentState === STATES.SHOWRESULT) {
      const result = this.model.state.currentResultIndex;
      const selected = this.model.state.currentUserSelectedIndex;
      if (result === selected) {
        this.model.state.score++;
        this.model.ui.roundResult = 'You Won!!';
      } else {
        this.model.ui.roundResult = 'Better Luck Next Time!!';
      }

      this.model.state.currentState = STATES.SELECTING;
      this.time = 0;
    }
  }
}
