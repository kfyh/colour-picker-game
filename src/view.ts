import { Container } from 'pixi.js';
import { Model, STATES } from './model';
import { GameView } from './view/gameView';
import { ResultView } from './view/resultView';

export class View {
  private readonly stage: Container;
  private readonly model: Model;

  private readonly gameContainer: Container;
  private readonly gameView: GameView;
  private readonly resultContainer: Container;
  private readonly resultView: ResultView;

  constructor(stage: Container, model: Model) {
    this.stage = stage;
    this.model = model;

    this.gameContainer = new Container();
    this.gameContainer.name = 'game';
    this.gameView = new GameView(this.gameContainer, this.model);

    this.resultContainer = new Container();
    this.resultContainer.name = 'result';
    this.resultView = new ResultView(this.resultContainer, this.model);
  }

  public buildScene(): void {
    this.gameView.buildScene();
    this.resultView.buildScene();

    this.resultContainer.visible = false;
    this.gameContainer.visible = true;

    this.stage.addChild(this.gameContainer);
    this.stage.addChild(this.resultContainer);
  }

  public update(delta: number): void {
    if (this.model.state.currentState === STATES.ENDED) {
      this.resultContainer.visible = true;
      this.gameContainer.visible = false;
      this.resultView.update(delta);
    } else {
      this.resultContainer.visible = false;
      this.gameContainer.visible = true;
      this.gameView.update(delta);
    }
  }

  public on(
    event: string,
    listener: (...args: any[]) => void,
    context?: any
  ): void {
    this.gameView.on(event, listener, context);
    this.resultView.on(event, listener, context);
  }

  public off(event: string, listener: (...args: any[]) => void): void {
    this.gameView.off(event, listener);
    this.resultView.off(event, listener);
  }
}
