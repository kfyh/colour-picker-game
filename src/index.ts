import { Application } from 'pixi.js';
import { Game } from './game';
import { Model, STATES } from './model';
import { View } from './view';


const gameModel: Model = {
    state: {
        currentResultIndex: 0,
        nextResultIndex: 0,
        currentUserSelectedIndex: 0,
        currentState: STATES.STARTED,
    },
    settings: {
        colours: [
            0x0000ff,
            0x00ff00,
            0xff0000,
            0xffff00,
            0xffa500
        ]
    }
}

const app = new Application({
    width: 400,
    height: 600,
    backgroundColor: 0xdcdcdc
});
document.body.appendChild(app.view);

const view = new View(app.stage, gameModel);
view.buildScene();

const game = new Game(gameModel);

let currentTime: number = -1
const update = (timeStamp: number): void => {
    let delta = 0;
    if (currentTime < 0) {
        currentTime = timeStamp;
    } else {
        delta = timeStamp - currentTime;
        currentTime = timeStamp;
    }

    game.update(delta);
    view.update(delta);

    requestAnimationFrame(update);
}

requestAnimationFrame(update);