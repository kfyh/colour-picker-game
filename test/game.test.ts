import { Game } from "../src/game";
import { Model, STATES } from "../src/model";

describe('game', () => {
    test('when game started, selection will start game', () => {
        const model: Partial<Model> = {
            state: {
                currentResultIndex: -1,
                currentUserSelectedIndex: -1,
                currentState: STATES.STARTED,
                timeRemaining: -1,
                score: -1
            },
            ui: {
                roundResult: ''
            }
        };
        const game = new Game(model as Model);
        game.setSelected(2);
        expect(model.state?.currentResultIndex).not.toBe(-1);
        expect(model.state?.currentUserSelectedIndex).toBe(2);
        expect(model.state?.currentState).toBe(STATES.CYCLING);
    });
});