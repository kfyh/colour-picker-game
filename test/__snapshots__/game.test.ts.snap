// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game after game started, when time is up then game will end 1`] = `
{
  "roundResult": "Time's Up!",
}
`;

exports[`game when cycling colour, after 2 seconds result is shown - loss 1`] = `
{
  "roundResult": "Better Luck Next Time!!",
}
`;

exports[`game when cycling colour, after 2 seconds result is shown - win 1`] = `
{
  "roundResult": "You Won!!",
}
`;
